(REF) CiviUnitTestCase - Plug more leaks

totten opened the pull request:
Overview
----------------------------------------

Fix some issues with setup/teardown. These came up when running locally with the sloppy-test detection.

This branch focuses on generic changes to CiviUnitTestCase which provide cleaner post-conditions across many tests. It will be interesting to see if the suites actually pass...

totten committed:
CiviUnitTestCase - renameLabels is outside txn, so resetLabels should be too

Context:

* CiviUnitTestCase::setUp() always runs before the optional call to `useTransaction()`.
* CiviUnitTestCase::setUp() calls renameLabels() to munge data. This outside/before the transaction.
* CiviUnitTestCase::tearDown() calls resetLabels() to un-munge data.

The problem here is that:

* renameLabels run OUTSIDE the transaction
* resetLabels runs INSIDE the transaction

So if you have transactions, then the reset will be rolled back... leaving
you with munged labels.

Here's a mnemonic to explain the change:

* What starts outside, ends outside
* What starts inside, ends inside

totten committed:
CiviUnitTestCase - Automatically restore settings

The base-class already takes a snapshot of your settings at the start, and
it provides a helper to restore settings 1-by-1.

This expands on that notion and aims to do it automatically.

colemanw commented:
@totten Jenkins says no

demeritcowboy commented on tests/phpunit/CiviTest/CiviUnitTestCase.php:
```diff
@@ -605,7 +609,12 @@ protected function tearDown(): void {
    */
   protected function revertSetting(string $setting): void {
-    \Civi::settings()->set($setting, $this->originalSettings[$setting]);
+    if (isset($this->settings[$setting])) {
```
```suggestion
    if (isset($this->originalSettings[$setting])) {
```

totten replied:
Doh! Thanks.

Included in the rebased version, which I'll push after a couple more fixes.

totten committed:
Settings - When listening for changes, expect possibility of "revert"

totten committed:
CiviUnitTestCase (etal) - Clear $permissions more realistically

The default value of `userPermissionClass->permissions` is NULL.

These calls to `unset()` actually delete the property in a way that's hard to inspect.

If we want to reset back to empty, then we can set to NULL -- which matches the default value.

totten committed:
CRM_Core_Permission_UnitTests - Preserve temp-perms when toggling extensions/components

totten committed:
ActivityTest - Add missing call to enable CiviCampaign

It didn't work on its own. And with stricter cleanups, it also wouldn't pass in a suite.

totten committed:
UserDashBoardTest - Set expected setting explicitly

This was passing before because the setting was provided by
testDashboardContentContributionsWithInvoicingEnabled() and
then leaked for use by testDashboardPartialPayments().

Without this declaration, the test fails in isolation -- and it also fails
when there's a stricter cleanup regime.

totten committed:
PDFLetterCommonTest - Ensure that hooks are registered after enabling CiviCampaign

In testAllContributionTokens(), it calls out to some helpers which
indirectly try to activate CiviCampaign.

* If CiviCampaign is already active, then it works. (This is what has traditionally
  happened when doing a full-suite test-run.)
* If CiviCampaign is inactive, then it fails. (This is what would happen when
  running individually - or when a stricter cleanup regime is in place.)

totten commented:
Rebased.

A lot of the test-failures were actually problematic already -- in the sense that they _passed as part of the suite_ but then _failed individually_. 

The tighter setup/teardown in this PR meant that the full-suite became more representative of individual-tests.

So the last few commits are fixing quirks in random tests.
